File: code_snippets/java/src/main/java/develop/SideEffects.java
//package develop;
//
//import dev.restate.sdk.Awaitable;
//import dev.restate.sdk.Awakeable;
//import dev.restate.sdk.ObjectContext;
//import dev.restate.sdk.common.CoreSerdes;
//import static greeter.generated.GreeterProto.*;
//
//import dev.restate.sdk.common.TerminalException;
//import greeter.generated.GreeterRestate;
//import org.apache.logging.log4j.core.Core;
//
//import java.time.Duration;
//import java.util.UUID;
//
//public class SideEffects extends GreeterRestate.GreeterRestateImplBase {
//
//    @Override
//    public GreetResponse greet(ObjectContext ctx, GreetRequest request) {
//
//        // <start_side_effect>
//        String output = ctx.sideEffect(CoreSerdes.JSON_STRING, () -> doDbRequest());
//        // <end_side_effect>
//
//
//        var paymentClient = new PaymentClient();
//        String txId = "";
//        int amount = 1;
//        // <start_retry_settings>
//        ctx.sideEffect(CoreSerdes.JSON_BOOLEAN, () -> {
//            boolean result = paymentClient.call(txId, amount);
//            if(result){
//                // highlight-next-line
//                throw new IllegalStateException("Payment failed");
//            } else {
//                return result;
//            }
//        });
//        // <end_retry_settings>
//
//
//        // <start_terminal>
//        try {
//            ctx.sideEffect(CoreSerdes.JSON_BOOLEAN, () -> {
//                boolean result = paymentClient.call(txId, amount);
//                if(result){
//                    //highlight-next-line
//                    throw new TerminalException(TerminalException.Code.UNKNOWN, "Payment failed");
//                } else {
//                    return result;
//                }
//            });
//        } catch (TerminalException e) {
//            // handle terminal error
//        }
//        // <end_terminal>
//
//
//        Awakeable<Boolean> a1 = ctx.awakeable(CoreSerdes.JSON_BOOLEAN);
//        Awakeable<Boolean> a2 = ctx.awakeable(CoreSerdes.JSON_BOOLEAN);
//        Awakeable<Boolean> a3 = ctx.awakeable(CoreSerdes.JSON_BOOLEAN);
//
//        // <start_combine_all>
//        Awaitable.all(a1, a2, a3).await();
//        // <end_combine_all>
//
//        // <start_combine_any>
//        boolean res = (boolean) Awaitable.any(a1, a2, a3).await();
//        // <end_combine_any>
//
//        // <start_uuid>
//        UUID uuid = ctx.random().nextUUID();
//        // <end_uuid>
//
//        // <start_random_nb>
//        int value = ctx.random().nextInt();
//        // <end_random_nb>
//
//
//
//        return GreetResponse.newBuilder()
//                .setMessage("Hello")
//                .build();
//    }
//
//    private String doDbRequest(){ return ""; }
//}
//
//class PaymentClient {
//    public boolean call(String txId, int amount) {
//        return true;
//    }
//}

File: code_snippets/kotlin/src/main/kotlin/develop/SideEffects.kt
package develop

import dev.restate.sdk.common.TerminalException
import dev.restate.sdk.kotlin.*
import java.util.*
import kotlin.time.Duration.Companion.seconds

internal class SideEffects {
    suspend fun sideEffect(ctx: Context) {
        // <start_side_effect>
        val output: String = ctx.runBlock { doDbRequest() }
        // <end_side_effect>

        val paymentClient = PaymentClient()
        val txId = ""
        val amount = 1

        // <start_retry_settings>
        ctx.runBlock {
            val result = paymentClient.call(txId, amount)
            if (result) {
                // withClass highlight-line
                throw IllegalStateException("Payment failed")
            } else {
                result
            }
        }
        // <end_retry_settings>


        // <start_terminal>
        try {
            ctx.runBlock {
                val result = paymentClient.call(txId, amount)
                if (result) {
                    // withClass highlight-line
                    throw TerminalException(TerminalException.INTERNAL_SERVER_ERROR_CODE, "Payment failed")
                } else {
                    result
                }
            }
        } catch (e: TerminalException) {
            // handle terminal error
        }
        // <end_terminal>

        val a1 = ctx.awakeable<Boolean>()
        val a2 = ctx.awakeable<Boolean>()
        val a3 = ctx.awakeable<Boolean>()

        // <start_combine_all>
        listOf(a1, a2, a3).awaitAll()
        // <end_combine_all>

        // <start_combine_any>
        val res = Awaitable.any(a1, a2, a3)
            .await() as Boolean

        // Or using the select statement
        val resSelect = select {
            a1.onAwait { it }
            a2.onAwait { it }
            a3.onAwait { it }
        }
        // <end_combine_any>

        // <start_uuid>
        val uuid: UUID = ctx.random().nextUUID()
        // <end_uuid>

        // <start_random_nb>
        val value: Int = ctx.random().nextInt()
        // <end_random_nb>
    }

    private fun doDbRequest(): String {
        return ""
    }
}

internal class PaymentClient {
    fun call(txId: String?, amount: Int): Boolean {
        return true
    }
}